File: examples/php/src/QueueClient.php
<?php

declare(strict_types=1);

namespace LocoAware\Webhook;

/**
 * Stand-in queue client.
 *
 * TODO: swap for your real broker:
 *   - AWS SQS / SNS        aws/aws-sdk-php
 *   - RabbitMQ             php-amqplib/php-amqplib
 *   - beanstalkd           pda/pheanstalk
 *   - Redis Streams        predis/predis
 *   - Or: Symfony Messenger / Laravel Queue + an `inbox` DB table
 *
 * publish() must return once the message is DURABLY persisted. Do not ack
 * the webhook response before publish() returns — if the PHP process dies
 * in between you lose the message and LocoAware won't redeliver it.
 */
class QueueClient
{
    public function publish(string $topic, string $rawBody): void
    {
        // Real implementation: SqsClient::sendMessage, AMQPChannel::basic_publish, etc.
        error_log("[queue] publish → {$topic} (" . strlen($rawBody) . " bytes)");
    }
}

File: examples/ruby/queue_client.rb
require 'json'

# Stand-in queue client.
#
# TODO: swap for your real broker:
#   - AWS SQS / SNS      aws-sdk-sqs, aws-sdk-sns
#   - RabbitMQ           bunny
#   - Redis / Sidekiq    sidekiq
#   - Shoryuken (SQS)    shoryuken
#   - Or an `inbox` table drained by a Sidekiq/Shoryuken worker.
#
# #publish must return only once the message is durably persisted. Don't ack
# the HTTP request before #publish returns — if the process dies between the
# two, you lose the message (you already said 2xx to LocoAware).
class QueueClient
  QUEUES = Hash.new { |h, k| h[k] = Queue.new }

  def self.publish(topic, raw_body)
    QUEUES[topic] << raw_body
  end

  def self.subscribe(topic, &handler)
    Thread.new do
      loop do
        body = QUEUES[topic].pop
        begin
          handler.call(JSON.parse(body))
        rescue => e
          warn "[worker] #{topic} handler failed: #{e.message}"
          # With a real broker, don't ack — message returns to the queue / DLQ.
        end
      end
    end
  end
end
